src/scipp/core/variable.py

```diff
@@ -385,10 +385,36 @@ def logspace(dim: str,
                  dtype=dtype)
 
 
+def _normalize_args(*, unit, **kwargs):
+    args = {**kwargs}
+    is_var = {
+        key: isinstance(val, _cpp.Variable)
+        for key, val in args.items() if val is not None
+    }
+    if any(is_var.values()):
+        if not all(is_var.values()):
+            arg_types = {key: type(val) for key, val in args.items()}
+            raise TypeError('Either arguments all or none have to be variables, '
+                            f'got types {arg_types}')
+        if unit == default_unit:
+            units = {key: val.unit for key, val in args.items() if val is not None}
+            if len(set(units.values())) != 1:
+                raise _cpp.UnitError(
+                    f'All units must be equal, got {units}. '
+                    'You can specify a unit explicitly with the `unit` argument.')
+            unit = next(iter(units.values()))
```

Comment:
Why are the units not checked if a unit is given?

Reply:
Because the given unit overrides the others and `to_unit` ensures that they are compatible.

Reply:
No test though?

Reply:
These two partially test it: https://github.com/scipp/scipp/blob/8ec883a7021f7df4895ea76256964272cb325d2e/tests/variable_creation_test.py#L452-L456

But if we follow through with this, there need to be more tests to probe more cases. I wanted to keep it small for now so i don't commit to something that gets discarded.